src/bin/pg_autoctl/primary_standby.c

```diff
@@ -1036,6 +1043,209 @@ primary_rewind_to_standby(LocalPostgresServer *postgres)
+				  "Postgres is not stopped");
+		return false;
+	}
+
+	/*
+	 * pg_rewind fails when the target cluster (meaning the local Postgres
+	 * instance) is either running or has not been shutdown correctly. Time to
+	 * use pg_controldata and see if the DBState there is to pg_rewind liking.
+	 */
+	const bool missingPgdataIsOk = false;
+
+	if (!pg_controldata(pgSetup, missingPgdataIsOk))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/*
+	 * We know that Postgres is not running thanks to pg_ctl_status, and we
+	 * just grabbed the output from pg_controldata. We can now implement the
+	 * same pre-condition checks as in Postgres pg_rewind.c.
+	 */
+	if (pgSetup->control.state != DB_SHUTDOWNED &&
+		pgSetup->control.state != DB_SHUTDOWNED_IN_RECOVERY)
+	{
+		/*
+		 * Before calling pg_rewind, attempt crash recovery on the Postgres
+		 * instance and then shutdown.
+		 */
+		ReplicationSource crashRecoveryReplicationSource = { 0 };
+
+		log_info("Postgres needs to enter crash recovery before pg_rewind.");
+
+		crashRecoveryReplicationSource = *replicationSource;
+
+		/* we target the earlier consistent state possible, or 'immediate' */
+		strlcpy(crashRecoveryReplicationSource.targetLSN,
+				"immediate",
+				sizeof(crashRecoveryReplicationSource.targetLSN));
+
+		/* pause when reaching target to avoid creating a new local timeline */
+		strlcpy(crashRecoveryReplicationSource.targetAction,
+				"pause",
+				sizeof(crashRecoveryReplicationSource.targetAction));
+
+		strlcpy(crashRecoveryReplicationSource.targetTimeline,
+				"current",
+				sizeof(crashRecoveryReplicationSource.targetTimeline));
+
+		if (!pg_setup_standby_mode(pgSetup->control.pg_control_version,
+								   pgSetup->pgdata,
+								   pgSetup->pg_ctl,
+								   &crashRecoveryReplicationSource))
+		{
+			log_error("Failed to setup for crash recovery "
+					  "in preparation for pg_rewind");
+			return false;
+		}
+
+		/*
+		 * Now that the configuration file is ready and asks for Postgres
+		 * shutdown when reaching crash recovery time, we start postgres as a
+		 * sub-process here and wait for it to terminate.
+		 */
+		fflush(stdout);
+		fflush(stderr);
+
+		/* time to create the node_active sub-process */
+		pid_t fpid = fork();
+
+		switch (fpid)
+		{
+			case -1:
+			{
+				log_error("Failed to fork the postgres supervisor process");
+				return false;
+			}
+
+			case 0:
+			{
+				/* execv() the postgres binary directly, as a sub-process */
+				(void) pg_ctl_postgres(pgSetup->pg_ctl,
+									   pgSetup->pgdata,
+									   pgSetup->pgport,
+
+				                       /* do not open the service just yet */
+									   pgSetup->listen_addresses,
+									   false);
+
+				/* unexpected */
+				log_fatal("BUG: returned from service_keeper_runprogram()");
+				exit(EXIT_CODE_INTERNAL_ERROR);
+			}
+
+			default:
+			{
+				/* wait until postgres crash recovery is done */
+				for (int attempts = 0;; attempts++)
+				{
+					int timeout = 30;
```

Comment:
These 30 seconds seem rather arbitrary, for a busy database this is probably not enough to do the recovery. I think it would be better if we would wait as long as the postgres process that we spawned is still running (and possibly making progress in the recovery).

Reply:
We have an infinite loop around a 30s timeout, so I believe we're ok here.